python/ruff-ecosystem/ruff_ecosystem/projects.py

```diff
@@ -26,14 +31,115 @@ class Project(Serializable):
     repo: Repository
     check_options: CheckOptions = field(default_factory=lambda: CheckOptions())
     format_options: FormatOptions = field(default_factory=lambda: FormatOptions())
+    config_overrides: ConfigOverrides = field(default_factory=lambda: ConfigOverrides())
 
     def with_preview_enabled(self: Self) -> Self:
         return type(self)(
             repo=self.repo,
             check_options=self.check_options.with_options(preview=True),
             format_options=self.format_options.with_options(preview=True),
+            config_overrides=self.config_overrides,
         )
 
+    def __post_init__(self):
+        # Convert bare dictionaries for `config_overrides` into the correct type
+        if isinstance(self.config_overrides, dict):
+            # Bypass the frozen attribute
+            object.__setattr__(
+                self, "config_overrides", ConfigOverrides(always=self.config_overrides)
+            )
+
+
+@dataclass(frozen=True)
+class ConfigOverrides(Serializable):
+    """
+    A collection of key, value pairs to override in the Ruff configuration file.
+
+    The key describes a member to override in the toml file; '.' may be used to indicate a
+    nested value e.g. `format.quote-style`.
+
+    If a Ruff configuration file does not exist and overrides are provided, it will be createad.
+    """
+
+    always: dict[str, Any] = field(default_factory=dict)
+    when_preview: dict[str, Any] = field(default_factory=dict)
+    when_no_preview: dict[str, Any] = field(default_factory=dict)
+
+    def __hash__(self) -> int:
+        # Avoid computing this hash repeatedly since this object is intended
+        # to be immutable and serializing to toml is not necessarily cheap
+        @cache
+        def as_string():
+            return tomli_w.dumps(
+                {
+                    "always": self.always,
+                    "when_preview": self.when_preview,
+                    "when_no_preview": self.when_no_preview,
+                }
+            )
+
+        return hash(as_string())
+
+    @contextlib.contextmanager
+    def patch_config(
+        self,
+        dirpath: Path,
+        preview: bool,
+    ) -> None:
+        """
+        Temporarily patch the Ruff configuration file in the given directory.
+        """
+        ruff_toml = dirpath / "ruff.toml"
+        pyproject_toml = dirpath / "pyproject.toml"
+
+        # Prefer `ruff.toml` over `pyproject.toml`
+        if ruff_toml.exists():
+            path = ruff_toml
+            base = []
+        else:
+            path = pyproject_toml
+            base = ["tool", "ruff"]
+
+        overrides = {
+            **self.always,
+            **(self.when_preview if preview else self.when_no_preview),
+        }
+
+        if not overrides:
+            yield
+            return
```

Comment:
I'm sure this is correct, but it reads strangely. Why do we need to `yield` (nothing) right before returning?

Reply:
If we don't yield, the function is no longer a generator and the caller will complain.

```python
from contextlib import contextmanager


@contextmanager
def foo():
    return


with foo():
    pass
```

```
Traceback (most recent call last):
  File "/Users/mz/eng/src/astral-sh/ruff/example.py", line 9, in <module>
    with foo():
  File "/Users/mz/.pyenv/versions/3.12.0/lib/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
TypeError: 'NoneType' object is not an iterator
```